### PR TITLE
Foreign app access

### PR DESCRIPTION
**Pull request by sczembor**
Closes #x

## Proposed Changes

- Add additional test scenarios for foreign app account access
- Allow to access foreign app account access bases on the appID

## TODO

- [x] update change log

### CODE REVIEW

File: packages/runtime/src/interpreter/interpreter.ts [hunk 1]
```diff
@@ -215,31 +215,50 @@ export class Interpreter {
...
+			return this._getAccountFromReference(accountRef, line, create);
 		} else {
 			return this._getAccountFromAddr(accountRef, line, create, immutable);
 		}
 
+	}
+	
+	/**
+	 * Queries accesible accounts by smart contract. It can be either accounts defined
+	 * in accounts field or accounts associated with apps decalred in foreingApps array
```
Comment (Megha-Dev-19) [inline packages/runtime/src/interpreter/interpreter.ts:233]:
```suggestion
	 * in accounts field or accounts associated with apps declared in foreingApps array
```

File: packages/runtime/src/interpreter/interpreter.ts [hunk 2]
```diff
@@ -215,31 +215,50 @@ export class Interpreter {
...
+	 * @param accountRef index of account or app ID
+	 * @param line line number
+	 * @param create create flag
```
Comment (Megha-Dev-19) [inline packages/runtime/src/interpreter/interpreter.ts:236]: may be we can rename it to `createFlag`.?

File: packages/runtime/test/src/interpreter/interpreter.ts [hunk 3]
```diff
@@ -198,6 +198,7 @@ describe("Interpreter", function () {
 				setUpInterpreter(7, 1e9);
 			});
 			it("Should not throw an error when accessing account that are in transaction's account field", function () {
+				// in which exact part are we accesing the account???
```
Comment (Megha-Dev-19) [inline packages/runtime/test/src/interpreter/interpreter.ts:201]: should it be there..?

Comment (sczembor) [reply packages/runtime/test/src/interpreter/interpreter.ts:201]: you are right, I just could not figure out what is going on here and left it there to ask the rest of the team. @thdailong could you provide some info about it?

File: packages/runtime/src/interpreter/interpreter.ts (hunk 2 above)
Comment (sczembor) [reply packages/runtime/src/interpreter/interpreter.ts:236]: that's a good idea, I will update it

File: packages/runtime/src/interpreter/interpreter.ts [hunk 4]
```diff
@@ -215,31 +215,50 @@ export class Interpreter {
...
-					? this.createAccountIfAbsent(address)
-					: this.runtime.ctx.state.accounts.get(address);
-			}
+
+		if (accountRef === 0n) {
+			address = encodeAddress(this.runtime.ctx.tx.snd);
+			account = this.runtime.ctx.state.accounts.get(address);
+			return this.runtime.assertAccountDefined(address, account, line);
+		} else if (typeof accountRef === "bigint") {
+			return this._getAccountFromReference(accountRef, line, create);
```
Comment (rmeena840) [inline packages/runtime/src/interpreter/interpreter.ts:224]: You are revering the condition? Please check if this is really needed.

**Review by rmeena840 — COMMENTED** [review]
left comment

File: packages/runtime/src/interpreter/interpreter.ts (hunk 4 above)
Comment (sczembor) [reply packages/runtime/src/interpreter/interpreter.ts:224]: I am not sure which part are you referring to

File: packages/runtime/src/interpreter/interpreter.ts [hunk 5]
```diff
@@ -55,7 +55,7 @@ export class Interpreter {
 	length: number; // total length of 'compiled' TEAL code
 	// local stores for a transaction.
 	bytecblock: Uint8Array[];
-	intcblock: BigInt[];
+	intcblock: bigint[];
```
Comment (rmeena840) [inline packages/runtime/src/interpreter/interpreter.ts:58]: I don't think we have bigint in small case. Please check this.

**Review by rmeena840 — COMMENTED** [review]
left comments

File: packages/runtime/src/interpreter/interpreter.ts (hunk 4 above)
Comment (rmeena840) [reply packages/runtime/src/interpreter/interpreter.ts]: `typeof accountRef === "bigint"` when this is true we were doing something but we are doing the opposite.

File: packages/runtime/src/interpreter/interpreter.ts (hunk 5 above)
Comment (sczembor) [reply packages/runtime/src/interpreter/interpreter.ts:58]: bigint small case is the type and BigInt is the constructor of the type

Comment (sczembor) [reply packages/runtime/src/interpreter/interpreter.ts:58]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt

File: packages/runtime/src/interpreter/interpreter.ts (hunk 4 above)
Comment (sczembor) [reply packages/runtime/src/interpreter/interpreter.ts]: we are doing the same it was just moved to a separate method

File: packages/runtime/src/interpreter/interpreter.ts (hunk 5 above)
Comment (rmeena840) [reply packages/runtime/src/interpreter/interpreter.ts:58]: ok

**Review by rmeena840 — APPROVED** [review]
Approved

**Pull request merged**